rootston/cursor.c

```diff
@@ -291,15 +307,86 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 
 void roots_cursor_handle_motion(struct roots_cursor *cursor,
 		struct wlr_event_pointer_motion *event) {
-	wlr_cursor_move(cursor->cursor, event->device,
-			event->delta_x, event->delta_y);
+	double dx = event->delta_x;
+	double dy = event->delta_y;
+
+	if (cursor->active_constraint) {
+		struct roots_view *view = cursor->pointer_view->view;
+		assert(view);
+
+		double center_x = view->x + view->width / 2.;
+		double center_y = view->y + view->height / 2.;
+
+		double lx1 = cursor->cursor->x;
+		double ly1 = cursor->cursor->y;
+
+		double lx2 = lx1 + dx;
+		double ly2 = ly1 + dy;
+
+		// Optimization for most common case.
+		// This also makes sure that we don't encounter
+		// precision bugs in the most common case.
+		if (view->rotation == 0.0) {
+			double sx1 = lx1 - view->x;
+			double sy1 = ly1 - view->y;
+
+			double sx2 = lx2 - view->x;
+			double sy2 = ly2 - view->y;
+
+			double sx2_confined, sy2_confined;
+			if (!wlr_region_confine(&cursor->confine, sx1, sy1, sx2, sy2, &sx2_confined, &sy2_confined)) {
+				return;
+			}
+
+			dx = sx2_confined - sx1;
+			dy = sy2_confined - sy1;
+		} else {
+			assert(false);
+			double c = cos(view->rotation);
```

Comment:
This code seems duplicated at least 3 times in rootston

Reply:
Bump

* https://github.com/swaywm/wlroots/pull/852/files#diff-fdc0e2c92f2086c22c357b7879ff1f1cR605
* https://github.com/swaywm/wlroots/pull/852/files#diff-68466fb724e9b87dc6d581fde3289e7aR799